db update_one operation with $set and  $elemMatch with dictionary object raises unknown toplevel operator $elemMatch error

database collection has one document to start with
```python
original_doc = { "recordId": 1234, 
                 "app": [
                                      {"application": "AppName", "code": 1234, "property": "oldValue"}, 
                                      {"application": "AppName1", "code": 1235, "property": "oldValue1"}
                         ]
               }
db_client["collection"].insert_one(key, { "$set": { "app.$" : doc }})
```

attempt to update the sub document with $elemMatch 
```python
key = {"recordId": 1234,  "app": { "$elemMatch": { "application":"AppName", "code": 1234}}}

subdoc_update = {"application": "AppName", "code": 1234, "property": "newValue"}

db_client["collection"].update_one(key, { "$set": { "app.$" : subdoc_update }})
```
# Expected
 works with mongo db and updates matching sub document

# Actual
fails with mongo mock with below error in filtering.py. looks like filtering is getting a subspec query from collections.py and its not able to handle $elemMatch at top level

```python
self = <mongomock.filtering._Filterer object at 0x00000287FDC78910>
search_filter = {'$elemMatch': {'application': 'AppName', 'code': 1234}}
document = {"application": "AppName'', "code": 1234, "property": "newValue"}

    def apply(self, search_filter, document):
        if not isinstance(search_filter, dict):
            raise OperationFailure('the match filter must be an expression in an object')
    
        for key, search in iteritems(search_filter):
            # Top level operators.
            if key == '$comment':
                continue
            if key in LOGICAL_OPERATOR_MAP:
                if not search:
                    raise OperationFailure('BadValue $and/$or/$nor must be a nonempty array')
                if not LOGICAL_OPERATOR_MAP[key](document, search, self.apply):
                    return False
                continue
            if key == '$expr':
                parse_expression = self.parse_expression[0]
                if not parse_expression(search, document, ignore_missing_keys=True):
                    return False
                continue
            if key in _TOP_LEVEL_OPERATORS:
                raise NotImplementedError(
                    'The {} operator is not implemented in mongomock yet'.format(key))
            if key.startswith('$'):
>               raise OperationFailure('unknown top level operator: ' + key)
E               pymongo.errors.OperationFailure: unknown top level operator: $elemMatch
```
Note: in my original documents there are lot more elements in the subdoc, only listed the fields necessary to demonstrate the issue

--- a/mongomock/collection.py
+++ b/mongomock/collection.py
@@ -1198,9 +1198,9 @@ class Collection(object):
                     subspec = spec
                     for part in field_name_parts[:-1]:
                         if part == '$':
-                            subspec = subspec.get('$elemMatch', subspec)
+                            subspec_dollar = subspec.get('$elemMatch', subspec)
                             for item in current_doc:
-                                if filter_applies(subspec, item):
+                                if filter_applies(subspec_dollar, item):
                                     current_doc = item
                                     break
                             continue
@@ -1219,7 +1219,8 @@ class Collection(object):
                     subdocument = current_doc
                     if field_name_parts[-1] == '$' and isinstance(subdocument, list):
                         for i, doc in enumerate(subdocument):
-                            if filter_applies(subspec, doc):
+                            subspec_dollar = subspec.get('$elemMatch', subspec)
+                            if filter_applies(subspec_dollar, doc):
                                 subdocument[i] = v
                                 break
                         continue

--- a/tests/test__mongomock.py
+++ b/tests/test__mongomock.py
@@ -1606,6 +1606,15 @@ class MongoClientCollectionTest(_CollectionComparisonTest):
             {'$setOnInsert': {'data.$.age': 1}}, True)
         self.cmp.compare.find()
 
+    def test__set_dollar_operand(self):
+        self.cmp.do.delete_many({})
+        self.cmp.do.insert_one({'recordId': 1234, 'app': [
+            {'application': 'AppName', 'code': 1234, 'property': 'oldValue'},
+            {'application': 'AppName1', 'code': 1235, 'property': 'oldValue1'}]})
+        self.cmp.do.update_many(
+            {'app': {'$elemMatch': {'application': 'AppName', 'code': 1234}}},
+            {'$set': {'app.$': {'application': 'AppName', 'code': 1234, 'property': 'newValue'}}})
+
     def test__addToSet(self):
         self.cmp.do.delete_many({})
         self.cmp.do.insert_one({'name': 'bob'})

